File: android/src/main/java/com/margelo/nitro/razorpaynitro/HybridRazorpay.kt
package com.margelo.nitro.razorpaynitro

import android.app.Activity
import android.content.Intent
import com.facebook.react.bridge.ActivityEventListener
import com.margelo.nitro.NitroModules
import com.margelo.nitro.core.Promise
import com.razorpay.Checkout
import org.json.JSONObject

/**
 * HybridRazorpay - Native Android implementation of the Razorpay Nitro Module.
 * Uses RazorpayPaymentActivity to handle payment callbacks properly.
 * Supports external wallet selection callbacks.
 */
class HybridRazorpay : HybridRazorpaySpec(), ActivityEventListener {

    // Store the pending promise for callback resolution
    private var pendingPromise: Promise<String>? = null

    companion object {
        private const val TAG = "HybridRazorpay"
        private const val REQUEST_CODE_RAZORPAY = 1001
        
        // Static callback for external wallet selection (shared across instances)
        private var externalWalletCallback: ((String) -> Unit)? = null
        
        /**
         * Sets the external wallet callback from JS layer.
         */
        fun setExternalWalletCallback(callback: ((String) -> Unit)?) {
            externalWalletCallback = callback
        }
    }

    init {
        NitroModules.applicationContext?.addActivityEventListener(this)
    }

    /**
     * Opens the Razorpay checkout with the provided options.
     */
    override fun open(options: String): Promise<String> {
        // Check for existing payment in progress
        if (pendingPromise != null) {
            return Promise.rejected(
                Exception(createError("PAYMENT_IN_PROGRESS", "A payment is already in progress"))
            )
        }

        // Get current activity
        val activity = NitroModules.applicationContext?.currentActivity
        if (activity == null) {
            return Promise.rejected(
                Exception(createError("ACTIVITY_NOT_FOUND", "No activity available to present Razorpay checkout"))
            )
        }

        // Validate options JSON
        try {
            JSONObject(options)
        } catch (e: Exception) {
            return Promise.rejected(
                Exception(createError("INVALID_OPTIONS", "Failed to parse options: ${e.message}"))
            )
        }

        // Create promise and store for callback resolution
        val promise = Promise<String>()
        pendingPromise = promise

        // Launch RazorpayPaymentActivity
        try {
            val intent = Intent(activity, RazorpayPaymentActivity::class.java).apply {
                putExtra(RazorpayPaymentActivity.EXTRA_OPTIONS, options)
            }
            activity.startActivityForResult(intent, REQUEST_CODE_RAZORPAY)
        } catch (e: Exception) {
            pendingPromise = null
            return Promise.rejected(
                Exception(createError("CHECKOUT_ERROR", "Failed to start payment activity: ${e.message}"))
            )
        }

        return promise
    }

    /**
     * Called when an activity returns a result.
     */
    override fun onActivityResult(activity: Activity, requestCode: Int, resultCode: Int, data: Intent?) {
        if (requestCode != REQUEST_CODE_RAZORPAY) return
        
        when (resultCode) {
            RazorpayPaymentActivity.RESULT_PAYMENT_SUCCESS -> {
                val promise = pendingPromise ?: return
                pendingPromise = null
                
                val paymentId = data?.getStringExtra(RazorpayPaymentActivity.EXTRA_PAYMENT_ID) ?: ""
                val orderId = data?.getStringExtra(RazorpayPaymentActivity.EXTRA_ORDER_ID) ?: ""
                val signature = data?.getStringExtra(RazorpayPaymentActivity.EXTRA_SIGNATURE) ?: ""
                
                val result = JSONObject().apply {
                    put("razorpay_payment_id", paymentId)
                    put("razorpay_order_id", orderId)
                    put("razorpay_signature", signature)
                }
                
                promise.resolve(result.toString())
            }
            RazorpayPaymentActivity.RESULT_PAYMENT_ERROR -> {
                val promise = pendingPromise ?: return
                pendingPromise = null
                
                val errorCode = data?.getIntExtra(RazorpayPaymentActivity.EXTRA_ERROR_CODE, -1) ?: -1
                val errorDescription = data?.getStringExtra(RazorpayPaymentActivity.EXTRA_ERROR_DESCRIPTION) 
                    ?: "Payment failed or cancelled"
                
                promise.reject(Exception(createError(errorCode.toString(), errorDescription)))
            }
            RazorpayPaymentActivity.RESULT_EXTERNAL_WALLET -> {
                // External wallet selected - call the callback but don't resolve promise yet
                val walletData = data?.getStringExtra(RazorpayPaymentActivity.EXTRA_WALLET_DATA) ?: "{}"
                externalWalletCallback?.invoke(walletData)
                // Don't clear pendingPromise - payment flow continues after wallet selection
            }
            Activity.RESULT_CANCELED -> {
                val promise = pendingPromise ?: return
                pendingPromise = null
                promise.reject(Exception(createError("0", "Payment cancelled by user")))
            }
            else -> {
                val promise = pendingPromise ?: return
                pendingPromise = null
                promise.reject(Exception(createError("-1", "Unknown payment result")))
            }
        }
    }

    /**
     * Called when a new intent is received.
     */
    override fun onNewIntent(intent: Intent) {
        // Not used for Razorpay
    }

    /**
     * Creates a JSON error string with code and message fields.
     */
    private fun createError(code: String, message: String): String {
        return JSONObject().apply {
            put("code", code)
            put("message", message)
        }.toString()
    }
}

File: ios/HybridRazorpay.swift
import Foundation
import UIKit
import NitroModules
import Razorpay

/**
 * HybridRazorpay - Native iOS implementation of the Razorpay Nitro Module.
 * Implements RazorpayPaymentCompletionProtocol and ExternalWalletSelectionProtocol.
 */
final class HybridRazorpay: HybridRazorpaySpec_base, HybridRazorpaySpec_protocol {
    
    // MARK: - Properties
    
    /// Closure to resolve the Promise on payment success
    private var pendingResolver: ((String) -> Void)?
    
    /// Closure to reject the Promise on payment error
    private var pendingRejecter: ((any Error) -> Void)?
    
    /// Razorpay checkout instance
    private var razorpay: RazorpayCheckout?
    
    /// Static callback for external wallet selection (shared across instances)
    private static var externalWalletCallback: ((String) -> Void)?
    
    // MARK: - Top View Controller Helper
    
    /**
     * Finds the topmost UIViewController in the view hierarchy.
     */
    private func topViewController() -> UIViewController? {
        guard let windowScene = UIApplication.shared.connectedScenes
            .compactMap({ $0 as? UIWindowScene })
            .first(where: { $0.activationState == .foregroundActive }),
              let window = windowScene.windows.first(where: { $0.isKeyWindow }) else {
            return nil
        }
        
        var topController = window.rootViewController
        
        while let presented = topController?.presentedViewController {
            topController = presented
        }
        
        return topController
    }
    
    // MARK: - Open Method
    
    /**
     * Opens the Razorpay checkout with the provided options.
     * @param options JSON string containing RazorpayOptions
     * @returns Promise resolving to JSON string of payment success data
     */
    func open(options: String) throws -> Promise<String> {
        let promise = Promise<String>()
        
        // Check for existing payment in progress
        guard self.pendingResolver == nil else {
            promise.reject(withError: self.createNSError(code: "PAYMENT_IN_PROGRESS", message: "A payment is already in progress"))
            return promise
        }
        
        // Execute on main thread for UI operations
        DispatchQueue.main.async { [weak self] in
            guard let self = self else {
                promise.reject(withError: RuntimeError.error(withMessage: "Instance deallocated"))
                return
            }
            
            // Get current view controller
            guard let viewController = self.topViewController() else {
                promise.reject(withError: self.createNSError(code: "VIEW_CONTROLLER_NOT_FOUND", message: "No view controller available to present Razorpay checkout"))
                return
            }
            
            // Parse options JSON
            guard let data = options.data(using: .utf8),
                  let optionsDict = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
                promise.reject(withError: self.createNSError(code: "INVALID_OPTIONS", message: "Failed to parse options"))
                return
            }
            
            // Validate required key
            guard let key = optionsDict["key"] as? String else {
                promise.reject(withError: self.createNSError(code: "INVALID_OPTIONS", message: "Missing required 'key' in options"))
                return
            }
            
            // Store promise callbacks
            self.pendingResolver = { result in
                promise.resolve(withResult: result)
            }
            self.pendingRejecter = { error in
                promise.reject(withError: error)
            }
            
            // Initialize Razorpay checkout with data delegate (preserves order_id/signature)
            self.razorpay = RazorpayCheckout.initWithKey(key, andDelegateWithData: self)
            
            // Set external wallet delegate for wallet selection support
            self.razorpay?.setExternalWalletSelectionDelegate(self)
            
            self.razorpay?.open(optionsDict, displayController: viewController)
        }
        
        return promise
    }
    
    // MARK: - External Wallet Selection Registration
    
    /**
     * Registers a callback for external wallet selection events.
     * @param callback JSON string callback when user selects an external wallet
     */
    func setExternalWalletCallback(callback: @escaping (String) -> Void) {
        HybridRazorpay.externalWalletCallback = callback
    }
    
    /**
     * Clears the external wallet callback.
     */
    func clearExternalWalletCallback() {
        HybridRazorpay.externalWalletCallback = nil
    }
    
    // MARK: - Error Helpers
    
    private func createNSError(code: String, message: String) -> NSError {
        return NSError(
            domain: "RazorpayError",
            code: 0,
            userInfo: [
                "code": code,
                "message": message
            ]
        )
    }
    
    private func clearPending() {
        pendingResolver = nil
        pendingRejecter = nil
        razorpay = nil
    }
}

// MARK: - RazorpayPaymentCompletionProtocolWithData

extension HybridRazorpay: RazorpayPaymentCompletionProtocolWithData {
    
    func onPaymentSuccess(_ payment_id: String, andData response: [AnyHashable: Any]?) {
        var result: [String: Any] = [
            "razorpay_payment_id": payment_id
        ]
        
        // Include additional data from response if available
        if let response = response {
            if let orderId = response["razorpay_order_id"] as? String {
                result["razorpay_order_id"] = orderId
            }
            if let signature = response["razorpay_signature"] as? String {
                result["razorpay_signature"] = signature
            }
        }
        
        if let data = try? JSONSerialization.data(withJSONObject: result),
           let jsonString = String(data: data, encoding: .utf8) {
            pendingResolver?(jsonString)
        } else {
            pendingResolver?("{\"razorpay_payment_id\":\"\(payment_id)\"}")
        }
        
        clearPending()
    }
    
    func onPaymentError(_ code: Int32, description str: String, andData response: [AnyHashable: Any]?) {
        pendingRejecter?(createNSError(code: String(code), message: str))
        clearPending()
    }
}

// MARK: - ExternalWalletSelectionProtocol

extension HybridRazorpay: ExternalWalletSelectionProtocol {
    
    func onExternalWalletSelected(_ walletName: String, withPaymentData paymentData: [AnyHashable: Any]?) {
        var walletInfo: [String: Any] = [
            "external_wallet": walletName
        ]
        
        // Include payment data if available
        if let paymentData = paymentData {
            for (key, value) in paymentData {
                if let stringKey = key as? String {
                    walletInfo[stringKey] = value
                }
            }
        }
        
        // Notify via static callback
        if let callback = HybridRazorpay.externalWalletCallback,
           let data = try? JSONSerialization.data(withJSONObject: walletInfo),
           let jsonString = String(data: data, encoding: .utf8) {
            callback(jsonString)
        }
    }
}

